### ISSUE TITLE
API Endpoint /users PUT /:uid Not Honoring Supplied uid

### ISSUE THREAD

**Issue by sotarules** [NodeBB/nodebb-plugin-write-api#23]
Friends,

I've successfully created new users using this API, and I've moved on to trying to update them as well.  

BTW, broadly speaking, I'm programmatically synchronizing NodeBB users with our own user database so that our administrators won't have to do "dual" maintenance as new users are added to the system.

What I'm seeing is that the API end point for updating a user record **/users PUT /:uid** is updating _my_ user record.  If I understand the documentation, my guess is that my uid 17 specified on the URL is being ignored.

Here is a log diagnostic from my system:

```
user_editor.js syncUserNodeBB /users PUT email=nodebb@pract.us uid=17 url=http://support.pract.us/api/v1/users/17 updateRequest={"headers":{"Authorization":"Bearer 7962b4a4-504e-41f0-b669-c7a02c696f27"},"data":{"username":"Testy E Coyote","fullname":"Testy E Coyote"}}
```
BTW I've changed some of the characters in the access token for security reasons.

When I do this, the system does update _something_ namely my own record.  Note: I happen to be uid 1.

![image](https://cloud.githubusercontent.com/assets/5454239/7442672/77e1d8ea-f0cf-11e4-932f-001fbf7e3b09.png)

Note that uid 17 "Wile E Coyote" remains unaffected; however, _my_ record uid 1 (previously "David Lynch") has been updated to "Testy E Coyote".

Does it look like my URL is correct or could there be some other considerations?

TIA,
Dave

**Comment by julianlam** [comment]
Please try with v3.0.5 of this plugin.

**Comment by sotarules** [comment]
Request:
```
curl --request PUT -H 'Authorization: Bearer 7962b4e4-504a-41f0-b669-f7b02c696f27' --data 'username=Worked&fullname=Worked' http://support.pract.us/api/v1/users/17
```

Response:
```
{"code":"ok","payload":{}}
```
As before, my user record (uid 1) is updated:

![image](https://cloud.githubusercontent.com/assets/5454239/7442942/70997ea0-f0dc-11e4-8f50-8e1d60bbd676.png)

Note in this case, I'm using a _user token_ that is associated with my user record.  I understand this is OK as long as I supply the proper uid on the URL.  If you want, I can also try this with a _master token_ and see if there is any effect.

**Comment by julianlam** [comment]
Please do, specify either 17 or 1 in the _uid argument.

**Comment by sotarules** [comment]
Julian, 3.0.5 seems to work, I just tried it with CURL to update uid 17 username to Zorked using a _user token_ and here is the result:

![image](https://cloud.githubusercontent.com/assets/5454239/7443049/6e4b18ae-f0e4-11e4-8c7a-e46faf52ff00.png)

Thank for helping me so attentively.

I have a question about something else that has to do with the standard, built-in read-only API, and I thought you might have some perspective on this.  When I do a GET to retrieve information about a particular uid, I get a lot of information, but one thing that is lacking is the email address as shown here:

![image](https://cloud.githubusercontent.com/assets/5454239/7443052/ded77a0e-f0e4-11e4-96b5-f5121098d434.png)

Do you know if this is _as designed_ or is this a bug?  If I could get the email address, I could close the loop on this synchronization feature.

Finally, is there any way (either now or planned) to allow the API to update the user images?  In my case, just being able to set the URL would do it, because the images are hosted on Amazon S3.  If this were possible, I could share the same user images between our system and NodeBB.

TIA for your help.

**Comment by julianlam** [comment]
No problem, glad to help.

This is designed. If the user elects not to share their email publicly,
then it will not appear in the read API.

Perhaps if the calling user is an admin then we will show it regardless,
what do you think?

I am worried about making behaviours too different based on logged in user,
so this may not be something we can implement right away.

Would be nice to have a "browse forum as" feature, but that is outside the
scope of this issue.

**Comment by sotarules** [comment]
Julian, for my type of application, I could benefit from carte blanche access to the information.  We're treating NodeBB as an extension of our main system and we're trying to "blur the lines" between the two. Our server will be doing updates to NodeBB behind the scenes.  I do understand the security concerns, but in a case like ours, with a powerful administrative user ID, I would recommend the blinders come off.

BTW, do you have any plans regarding allowing user images (i.e., field **picture**) to be updated via the API?

**Comment by julianlam** [comment]
@barisusakli Was I right in assuming that email is hidden based on preferences? Couldn't find it after I took a cursory look through the code, so let me know if you think exposing it for admins is doable.

**Comment by sotarules** [comment]
I'll do some tests Monday and post what I learn here about the preferences.

Question on the user create (POST) transaction.  I'm seeing that when I supply a username that already exists, the system is assigning a seemingly random number and appending it to the name (the transaction succeeds).

Is this assignment of a number to suffix the user name by-design (i.e., can I rely on that behavior as a feature)?  Is it a random number or is there some algorithm that determines the value?

**Comment by julianlam** [comment]
It's by design, and it's a random three-digit number. We still reject registrations of usernames that already exist, but user creations that are automated (such as via write-api or through OAuth SSO) cannot handle rejections, so we altered the behaviour so that a random number is appended.

**Comment by sotarules** [comment]
Julian I did some tests today.  First thing, I started using a _master token_, and that seems to be working just as advertised, nothing newsworthy there.

I also confirmed that the privacy settings are in fact controlling the responses from the standard user API GET transaction:

![image](https://cloud.githubusercontent.com/assets/5454239/7447730/6c9bb194-f1bb-11e4-9ddc-b3fcdd40df71.png)

Once I check these two boxes, my email starts appearing.  When I use this URL:

http://support.pract.us/api/user/uid/1

I get this:

![image](https://cloud.githubusercontent.com/assets/5454239/7447733/85fdac78-f1bb-11e4-8d6c-83deb27cfd75.png)

So you were right about the privacy settings.
